File: src/main/java/pm/n2/tangerine/KeyboardManager.java
package pm.n2.tangerine;

import org.quiltmc.qsl.base.api.event.Event;

import java.util.ArrayList;

public class KeyboardManager {
	public final Event<KeyPress> KEY_PRESS = Event.create(KeyPress.class, listeners -> key -> {
		for (KeyPress listener : listeners) listener.onKeyPress(key);
	});

	private final ArrayList<Integer> pressedKeys = new ArrayList<>();

	public void keyPress(int key) {
		if (!pressedKeys.contains(key)) {
			pressedKeys.add(key);
			KEY_PRESS.invoker().onKeyPress(key);
		}
	}

	public void keyRelease(int key) {
		if (pressedKeys.contains(key)) pressedKeys.remove((Integer) key);
	}

	public boolean isKeyPressed(int key) {
		return pressedKeys.contains(key);
	}

	public interface KeyPress {
		void onKeyPress(int key);
	}
}

File: src/main/kotlin/pm/n2/tangerine/KeyboardManager.kt
package pm.n2.tangerine

import pm.n2.tangerine.core.TangerineEvent

object KeyboardManager {
    private val pressedKeys = ArrayList<Int>()

    fun isKeyPressed(key: Int): Boolean {
        return pressedKeys.contains(key)
    }

    fun registerKeyPress(key: Int): Boolean {
        if (!pressedKeys.contains(key)) {
            pressedKeys.add(key)
            return Tangerine.eventManager.dispatch(TangerineEvent.KeyPress(key)).filterIsInstance<Boolean>().any { it }
        }

        return false
    }

    fun registerKeyRelease(key: Int) {
        if (pressedKeys.contains(key)) pressedKeys.remove(key)
    }
}
